File: 프로그래머스/lv2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.js
function solution(elements) {
    const sumSet = new Set();
    const len = elements.length;

    for (let w = 1; w <= len; w++) { // w는 window의 크기 // w = 1, 2, 3, 4, 5 (창문 크기)
        let windowSum = 0;
        
        for (let i = 0; i < len; i++) { // 배열 순회 i = 0, 1, 2, 3, 4 어디서부터 더하느냐 (시작점)
            if (i === 0) { // 첫번째만 계산해줌
                for (let j = 0; j < w; j++) { // 창문 크기만큼 반복문으로 더하기 3 (창문개수만큼 시작점 시작해서 더한다)
                    windowSum += elements[j]; // 7
                }
            } else { // 그 뒤의 경우에는
                windowSum -= elements[i - 1]; // 7 - 7
                windowSum += elements[(i + w - 1) % len]; // 0 + 9
            }
            sumSet.add(windowSum); // {7, 9}
        }
    }
    
    return sumSet.size;
}

File: 프로그래머스/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.py
def solution(elements):
    sum_set = set()
    extended_elements = elements * 2
    
    # 길이별로 수열의 합 구하기
    for sub_len in range(1, len(elements) + 1): # 1, 2, 3, 4, 5
        for i in range(0, len(extended_elements)):
            sub_sum = extended_elements[i : i + sub_len]
            sum_set.add(sum(sub_sum))
            
    return len(sum_set)

# 위 방법은 시간 복잡도가 O(n^3)
# 시간 복잡도를 줄이기 위해 sub_sum을 구하는 부분에서 매번 슬라이싱하는 대신 누적 합을 사용할 수 있다.
# 누적 합 : 각 부분 수열의 첫 번째 합은 미리 계산하고, 그 후에는 한 칸씩 이동할 때 앞에서 하나를 빼고 뒤에서 하나를 더하는 방식으로 부분합을 업데이트하는 방법

# def solution(elements):
#     sum_set = set()
#     extended_elements = elements * 2
    
#     # 길이별로 수열의 합 구하기
#     for sub_len in range(1, len(elements) + 1): # 1, 2, 3, 4, 5
#         sub_sum = sum(elements[0 : sub_len]) # 초기 부분합 계산
#         sum_set.add(sub_sum)
        
#         # 한 칸씩 이동하며 부분합을 업데이트
#         for i in range(1, len(elements)):
#             sub_sum -= extended_elements[i - 1]
#             sub_sum += extended_elements[i + sub_len - 1]
#             sum_set.add(sub_sum)
            
#     return len(sum_set)

# O(n^2)